wemake_python_styleguide/types.py

```diff
@@ -98,35 +98,93 @@ class or structure.
     """
 
     # General:
-    min_name_length: int
-    i_control_code: bool
-    max_name_length: int
-    max_noqa_comments: int
-    nested_classes_whitelist: List[str]
-    allowed_domain_names: List[str]
-    forbidden_domain_names: List[str]
+    @property
```

Comment:
Why do we need a `@property` decorator here?

Reply:
Because of following error:

```
wemake_python_styleguide/checker.py:139: error: Incompatible types in assignment (expression has type "_ValidatedOptions", variable has type "ConfigurationOptions")
wemake_python_styleguide/checker.py:139: note: Protocol member ConfigurationOptions.max_call_level expected settable variable, got read-only attribute
```

By default all protocol's attributes are settable and according to few issues on [`mypy` repo](https://github.com/python/mypy/issues/6002#issuecomment-550421647) or [`typing` repo](https://github.com/python/typing/issues/622) the only wayto define unsettable protocol's attribute is to use `@property` without defining `setter` and/or `deleter`.

Why `ConfigurationOptions` protocol was introduced?
Doesn't it possible to make `_ValidatedOptions` public and use it in all annotations instead?

Reply:
This is really strange. Why this error happens now? (I am asking because I cannot tell the reason from the code changes)

Maybe because of the `frozen=True`?

Reply:
Yes, it's because of `frozen=True`
I wrote some explanation in issue description, but in short, to use `lru_cache` any function's argument must be hashable (so also immutable, because otherwise it would be really inconsistent and error-prone) and I have used `frozen=True` to make `_ValidatedOptions` instances immutable and because protocol `ConfigurationOptions` is used as type annotation for `_ValidatedOptions` instances it had to be changed to be compatible with now immutable `_ValidatedOptions`

Reply:
Got it, thanks!